### PR TITLE
memory fix

### PR DESCRIPTION
**Pull request by parisholley**
this pull fixes memory leak issues with the UDP service

### CODE REVIEW

**Comment by otoolep** [comment]
Thanks @parisholley -- the code needs 'go fmt' run on it. https://circleci.com/gh/influxdb/influxdb/6303

Have you signed the CLA?

Also, it's not clear to me there was a leak. Are you saying there was? I agree that the code may have been allocating excessively, but that is not the same as a leak. Am I missing something? In your change you now allocate when processing the individual points. Is there any actual difference with regards to allocation?

File: models/points.go [hunk 1]
```diff
@@ -193,10 +193,18 @@ func parsePoint(buf []byte, defaultTime time.Time, precision string) (Point, err
 		return nil, err
 	}
 
+	// convert from slices to avoid buffer overwrites from services like UDP
+	keyBuf := make([]byte, len(key))
```
Comment (otoolep) [inline models/points.go]: Isn't this just introducing even more allocation? buf -> key -> keyBuf? Previously it was just buf -> key.

Comment (parisholley) [inline models/points.go]: I would imagine so, however it appears go is able to GC this without issue.

**Comment by parisholley** [comment]
@otoolep yep, CLA signed.

yes, you can check out the heap dumps from my previous ticket on #4285. without this change, i can consistently crash the server. digging through pprof showed the buffer just kept building over time. it is possible that the write timeouts could also have been a trigger, but even so, it didn't seem like GC was kicking in properly.

**Comment by otoolep** [comment]
OK, interesting. Can you also provide an memory pprof after your change?

**Comment by parisholley** [comment]
http://imgh.us/heap_1443909031.svg

with this change, i am only at about 4% memory usage on a 30GB AWS box. before, it was a minimum of 15% and would eventually rise until 85%+ then crash...

**Comment by pauldix** [comment]
@jwilder should look at this before it gets merged. Also, @parisholley just mentioned on irc that this might be causing some other problems.

**Comment by parisholley** [comment]
@pauldix @jwilder see comment on #4321, i removed this changed and thought the bad data writes went away but it has returned. though I wouldn't be surprised if the distinction between slices/arrays is biting me again else where in the new storage engine code.

**Comment by jwilder** [comment]
Hmm.  This adds 3 allocations to the parsing code reduces throughput by 5-10%.  That code is called in a lot of performance sensitive places.  I would like to understand better how this fixes a memory leak though.  I could see how buffers that are re-used across connections could cause problems, but I don't understand how this fixes a memory leak.  If writes were timing out, than point data could be held in memory somewhere so whatever is allocating the memory is going to show up in the heap profile.

In any event, I'd suggest doing:
```
diff --git a/models/points.go b/models/points.go
index 1ded723..e64bec7 100644
--- a/models/points.go
+++ b/models/points.go
@@ -163,7 +163,9 @@ func ParsePointsWithPrecision(buf []byte, defaultTime time.Time, precision strin

 }

-func parsePoint(buf []byte, defaultTime time.Time, precision string) (Point, error) {
+func parsePoint(b []byte, defaultTime time.Time, precision string) (Point, error) {
+       buf := make([]byte, len(b))
+       copy(buf, b)
```

vs doing the 3 separate allocations and copies.  Benchmarking it shows it performs just slightly better for me.  Probably because it only adds 1 allocation and copy vs 3.  I'd be fine merging that change since it would also prevent the possibility of shared buffer data corruptions.  Let me know if that also fixes your memory issue.  

Thanks for looking into this @parisholley!

**Comment by parisholley** [comment]
@jwilder I know it did make a positive impact for ongoing uptime (when writes were working), but you are right that the OOM I was hitting could very well have been the result of writes timing out and golang not properly GC'ing the buffer in UDP service.

will need to wait on new storage engine to test this and will update the pull after but shared buffer is probably a safe enough reason to pull.

**Comment by otoolep** [comment]
Thanks @jwilder -- as I mentioned, I too don't see how existing code could cause a leak per-se.

**Comment by otoolep** [comment]
Let me do some more work in this area. While we might be able to make this more efficient, I see no leak in this code and want to understand what is going on.

**Comment by parisholley** [comment]
@oltoolep sure thing, i just had my instance crash due to OOM so I am going to bump the instance size and try to capture some more heap dumps when it grows. it is likely my indexes are large due to the high cardinality of tags and not a memory issue from UDP server (which may have been caused by slow writes in past).

**Comment by otoolep** [comment]
I'm closing this, as there is no evidence we're leaing memory. We may come back to this.

Thanks for the contribution @parisholley

**Pull request closed without merging**